community/bolt/src/main/java/org/neo4j/bolt/transport/SocketTransportHandler.java

```diff
@@ -172,16 +189,23 @@ private void chooseProtocolVersion( ChannelHandlerContext ctx, ByteBuf buffer )
 
         /**
          * @param availableVersions version -> protocol mapping
+         * @param encryptionRequired whether or not the server allows only encrypted connections
+         * @param isEncrypted whether of not this connection is encrypted
          */
-        public ProtocolChooser( PrimitiveLongObjectMap<BiFunction<Channel,Boolean,BoltProtocol>> availableVersions, boolean isEncrypted )
+        public ProtocolChooser( Map<Long, BiFunction<Channel, Boolean, BoltProtocol>> availableVersions, boolean encryptionRequired, boolean isEncrypted )
         {
             this.availableVersions = availableVersions;
+            this.encryptionRequired = encryptionRequired;
             this.isEncrypted = isEncrypted;
         }
 
         public HandshakeOutcome handleVersionHandshakeChunk( ByteBuf buffer, Channel ch )
         {
-            if ( handShake.remaining() > buffer.readableBytes() )
+            if ( encryptionRequired && !isEncrypted )
+            {
+                return HandshakeOutcome.INSECURE_HANDSHAKE;
```

Comment:
Wondering if this will work. If we turned on encryption, will it by default read the input bytes for tls/ssl handshake? Then the server will have already failed at client tls/ssl hello before the message reach here. A test to verify this code would help a lot.